active/003-seq-feature.md

```diff
@@ -0,0 +1,108 @@
+
+```python
+    interval_metadata = IntervalMetadata()
+        interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+        interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+        interval_metadata.update(Feature(gene='sagB', cds=True),
+                                 Feature(gene='sagB', cds=False))
+```
+
+
+`add(feature, *intervals)`
+- `feature` : `skbio.Feature`. new feature object add into the `IntervalMetadata` object
+- `*intervals` : a list of tuples corresponding to intervals where the feature object exists.
+- This allows for a single feature (include those that have non-continugous features) to be added into the `IntervalMetadata` object.
+
+```python
+   interval_metadata = IntervalMetadata()
+   interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+   interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+```
+
+`query(*args, **kwargs)`
+- `*args` : a list of tuples to search for features
+- `**kwargs` : keyword arguments to query features by their attributes.
+- This allows for features to be searched by both intervals and keywords (i.e. CDS vs exon, ...)
+- Note that for a specific interval query, multiple features can be returned.
+ 
+```python
+   interval_metadata = IntervalMetadata(features={
+                                        Feature(gene='sagA', location=0): [(0, 2), (4, 7)],
+                                        Feature(gene='sagB', location=0): [(3, 5)]})
+ 
+   feats = interval_metadata.query((1, 2))
+   feats = interval_metadata.query(gene='sagB')
+```
+
+This is the current implementation in micronota and fits all of our use-cases at the moment.  We don't feel that IntervalIndex objects as suggested in the recent pandas poll request can easily handle all of our use cases (i.e. non-contiguous intervals isn't straightforward).
+
+##Drawbacks
+- The `query` method does a linear lookup when searching for attributes over features.  This problem can blow up (in terms of development) if we want faster querying.  Which is why we opted for slow, linear search.  Again suggestions here are welcome.
+- features and intervals are loosely coupled.
+
+#Unresolved Questions
+- The `update` method may not be the most straightforward for users.  The reason why we opted for this design was because the `skbio.Feature` object can only be indexed uniquely by hashing the entire feature.  Suggestions on how to improve this will be welcome.
+- Is there a better way to associate intervals to their corresponding features?
+- How will features be deleted?
+- How will intervals for a given feature be updated?
+- How exactly will slicing be handled?  May want to be able to pass intervals/features into `__getitem__` within the `skbio.Sequence` class
+- How exactly should features be added to the interval_metadata?
+- What IntervalTree data structure should we stick with?
```

Comment:
I agree that we shouldn't be implementing highly optimized interval tree data structures, but we still have these tradeoffs to consider when choosing an interval tree implementation:

- If we can use an existing interval tree *package* that is already maintained, that's ideal because we can simply start using it and the only cost is an extra skbio dependency.
- If we use an interval tree implementation that's not in a package (or we need to strip it from a package, e.g. bx-python), that implementation will need to live in scikit-bio or in a new package, and that code will need a maintainer in either case. If we go this route, we need to have a clear maintainer for that code. One way to identify a maintainer is to have them use an existing implementation as a reference (e.g., bx-python's) but rewrite it so that a scikit-bio developer understands the code.

Since there's a lot of discussion about different interval tree implementations, I'd opt to keep the interval tree data structure private, at least initially. That way we are free to swap out different implementations without affecting users, which seems important at least while the functionality is experimental.